src/sparse/impl/KokkosSparse_twostage_gauss_seidel_impl.hpp

```diff
@@ -0,0 +1,919 @@
+          entries2(entries2_)
+        {}
+
+        // for storing both L&U values 
+        TwostageGaussSeidel_functor (
+                  bool two_stage_,
+                  const_ordinal_t num_rows_,
+                  input_row_map_view_t  rowmap_view_,
+                  input_entries_view_t  column_view_,
+                  input_values_view_t   values_view_,
+                  output_row_map_view_t row_map_,
+                  output_values_view_t  values_,
+                  output_values_view_t  diags_,
+                  output_row_map_view_t row_map2_,
+                  output_values_view_t  values2_):
+          two_stage(two_stage_),
+          num_rows(num_rows_),
+          rowmap_view(rowmap_view_),
+          column_view(column_view_),
+          values_view(values_view_),
+          row_map(row_map_),
+          entries(),
+          values(values_),
+          diags(diags_),
+          row_map2(row_map2_),
+          entries2(),
+          values2(values2_)
+        {}
+
+        // for generating ptr with parallel-scan
+        TwostageGaussSeidel_functor (
+                  const_ordinal_t num_rows_,
+                  output_row_map_view_t row_map_,
+                  output_row_map_view_t row_cnt_):
+          two_stage(true),
+          num_rows(num_rows_),
+          rowmap_view(),
+          column_view(),
+          values_view(),
+          row_map(row_map_),
+          entries(),
+          values(),
+          diags(),
+          row_cnt(row_cnt_)
+        {}
+
+        // ------------------------------------------------------- //
+        // functor for counting nnzL (with parallel_reduce)
+        KOKKOS_INLINE_FUNCTION
+        void operator()(const Tag_countNnzL&, const ordinal_t i, ordinal_t &nnz) const
+        {
+          ordinal_t nnz_i = 0;
+          for (size_type k = rowmap_view (i); k < rowmap_view (i+1); k++) {
+            if (column_view (k) < i) {
+              nnz_i ++;
+            } else if(!two_stage && column_view (k) == i) {
+              nnz_i ++;
+            }
+          }
+          row_map (i+1) = nnz_i;
+          if (i == 0) {
+            row_map (0) = 0;
+          }
+          nnz +=  nnz_i;
+        }
+
+        // functor for storing entriesL (with parallel_for)
+        KOKKOS_INLINE_FUNCTION
+        void operator()(const Tag_entriesL&, const ordinal_t i) const
+        {
+          ordinal_t nnz = row_map (i);
+          for (size_type k = rowmap_view (i); k < rowmap_view (i+1); k++) {
+            if (column_view (k) < i) {
+              entries (nnz) = column_view (k);
+              nnz ++;
+            } else if(!two_stage && column_view (k) == i) {
+              entries (nnz) = column_view (k);
+              nnz ++;
+            }
+          }
+        }
+
+        // functor for storing valuesL (with parallel_for)
+        KOKKOS_INLINE_FUNCTION
+        void operator()(const Tag_valuesL&, const ordinal_t i) const
+        {
+          const_scalar_t one (1.0);
```

Comment:
You probably want to use `Kokkos::ArithTraits` for that kind of assignment in case of complex type or even stranger things like sacado, simd, etc...